This notebook illustrate how to connect to remote service.

## `Aletheia` service

`aletheia` service provides nice interface to read/write memory systems composed of hyper-vectors.

Currently it provides the following operations: 

* `Get`: get/retrieve chunks that meets certain criteria;

* `Set`: set/write chunks with certain metadata;

The chunks can be terminal, representing non-divisible concepts.
In addition, we can produce composite chunk that are sequences of chunks (terminal or composite), set of chunks, etc.

For example, if we model the English alphabet as terminal chunks, then each English word can be modelled as sequence of alphabets.

NOTE this requires `alatheia` service, either running at local machine or an Internet-accessible IP.


In [1]:
import grpc
from kongming.util import remote_memory
from kongming.data import lang

from kongming.api import chunk_pb2, memory_pb2 

with grpc.insecure_channel(target="localhost:8080") as channel:
    m = remote_memory.RemoteMemory(channel)
    lang.populate_ascii(m)
    lang.populate_words(m, "lang:en", [
        "the", "quick", "brown", "fox", "jumps", "over", "the", "lazy", "dog",
        "world", "catherine", "university", "rate", "kitchen",
        "something", "activity", "necessary", "other", "stuff",
        "pattern", "director", "dying", "chinese",
        "specialize", "some", "for", "cat",
        "relax", "ing", "lifetime",
        "uzbekistan",
        "statistical"])

The script above connects to `aletheia` service, write English alphabets, followed by constructing some English words (as sequences of alphabets). 
Fuzzy query can be performed like:

In [3]:
with grpc.insecure_channel(target="localhost:8080") as channel:
    m = remote_memory.RemoteMemory(channel)
    wordPicker = memory_pb2.ChunkPicker(
            ref=chunk_pb2.ChunkReference(
                namespace="lang:en",
                token="world",
            ))
    got = m.Get(memory_pb2.ChunkPicker(near_neighbor=memory_pb2.NearNeighbor(target=wordPicker)))
    print(got)

[elemental {
  hint: SPARSE_CONSTRAINED
  model: MODEL_1M_10BIT
  stable_hash: 4457954920353104723
  sparse_constrained {
    seed: 14336416972779868589
  }
}
category: SEQUENCE
refs {
  namespace: "lang:en"
  token: "world"
}
weight: 1024
, elemental {
  hint: SPARSE_CONSTRAINED
  model: MODEL_1M_10BIT
  stable_hash: 4334755355221515437
  sparse_constrained {
    seed: 13896323908361552193
  }
}
category: SEQUENCE
refs {
  namespace: "lang:en"
  token: "for"
}
weight: 148
, elemental {
  hint: SPARSE_CONSTRAINED
  model: MODEL_1M_10BIT
  stable_hash: 2914648003062445619
  sparse_constrained {
    seed: 3401615469385435847
  }
}
category: SEQUENCE
refs {
  namespace: "lang:en"
  token: "fox"
}
weight: 81
, elemental {
  hint: SPARSE_CONSTRAINED
  model: MODEL_1M_10BIT
  stable_hash: 5239658206539774533
  sparse_constrained {
    seed: 15671600646850579412
  }
}
category: SEQUENCE
refs {
  namespace: "lang:en"
  token: "dog"
}
weight: 68
, elemental {
  hint: SPARSE_CONSTRAINED
  model:

This result is worth careful examination and discussion.

* First of all, the word `world` itself is returned (as full match, overlap=1024);
* the word `for` is returned as second, with weight/overlap=148, since there are two letters overlap: `o` and `r`;
* the word `fox` is returned, with weight/overlap=81, 1 letter overlapping: `o`;
* the word `dog` is returned, with weight/overlap=68, 1 letter overlapping: `o`;
* the word `some` is returned, with weight/overlap=46, 1 letter overlapping: `o`;
* the word `director` is returned, with weight/overlap=29, 1 letter overlapping: `r`;
* the word `something` is returned, with weight/overlap=24, 1 letter overlappng: `o`.

All fuzzy matching is done at the server side, modelling cognitive similarity with the query word.
Currently the matching can be achieved within a few milli-seconds.

Similarly, spelling suggestions (for other words) also work perfectly.